codec: from_parts() for FramedWrite and FramedRead

Mrreadiness opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests.

Contributors guide: https://github.com/tokio-rs/tokio/blob/master/CONTRIBUTING.md

The contributors guide includes instructions for running rustfmt and building the
documentation, which requires special commands beyond `cargo fmt` and `cargo doc`.
-->
Implement `from_parts()` for `FramedWrite` and `FramedRead`.

## Motivation

Thanks to #7566, `FramedWrite` and `FramedRead` can be transformed into `FramedParts`, but there is no way to create them back from parts, or to transform `Framed` into `FramedRead` or `FramedWrite` without losing progress in buffers.

My original problem was splitting `Framed` into both `FramedRead` **and** `FramedWrite`, to swap the codec for one of them with `map_codec`. I wish we could have something like `into_split` for `Framed`:
```rust
impl<T, U> Framed<T, U>
where
    T: AsyncRead + AsyncWrite,
    U: Clone,
{
    pub fn into_split(self) -> (FramedRead<ReadHalf<T>, U>, FramedWrite<WriteHalf<T>, U>) {
        let (r, w) = split(self.inner.inner);
        let read = FramedRead::from_parts(FramedParts {
            io: r,
            codec: self.inner.codec.clone(),
            read_buf: self.inner.state.read.buffer,
            write_buf: BytesMut::new(),
            _priv: (),
        });
        let write = FramedWrite::from_parts(FramedParts {
            io: w,
            codec: self.inner.codec,
            read_buf: BytesMut::new(),
            write_buf: self.inner.state.write.buffer,
            _priv: (),
        });
        return (read, write);
    }
}

```
But cloning of the codec doesn't feel right here. So, with `from_parts`, a user will have more control over the reader and writer reconstruction.

## Solution
Implement `from_parts()` for `FramedWrite` and `FramedRead`, which builds those structures from `FramedParts` and preserves internal buffers.

ADD-SP reviewed (commented):
Would mind adding some tests for both `from_parts` and `into_parts`?

Mrreadiness commented:
@ADD-SP Thanks! I added tests as well.

I also noticed that `FramedParts::new()` forces the creation of parts with at least an `Encoder` codec. But it doesn't make much sense, because it could be used for `Decoder` without `Encoder`. Looks like it's just a historical mistake, and there was an attempt to remove this limitation (#5280), but it's a breaking change, so it was reverted. 
That's why I added an alternative constructor: `FramedParts::new_parts`. 

So, solving the task of splitting up Framed will look like:
```rust
let framed = Framed::from_parts(parts);
let FramedParts {
    io,
    read_buf,
    write_buf,
    ..
} = framed.into_parts();
let (io_read, io_write) = tokio::io::split(parts.io);
let mut read_parts = FramedParts::new_parts(io_read, Codec::new());
read_parts.read_buf = read_buf;
let framed_read = FramedRead::from_parts(read_parts);
let mut write_parts = FramedParts::new_parts(io_write, Codec::new());
write_parts.write_buf = write_buf;
let framed_write = FramedWrite::from_parts(read_parts);
```
It's far from perfect, but at least doable now. 
If you see a better way to do such a split or even an ergonomic way to add it as a method of `Framed`, I will be happy to implement it.

Mrreadiness commented:
Hi @ADD-SP, @martin-g!
Do you have any ideas how to improve the pr?

ADD-SP commented on tokio-util/src/codec/framed_read.rs:
```diff
@@ -161,6 +161,18 @@ impl<T, D> FramedRead<T, D> {
+            inner: FramedImpl {
+                inner: parts.io,
+                codec: parts.codec,
+                state: parts.read_buf.into(),
```
@conradludgate What do you think?

conradludgate replied:
I think it would make sense to add those fields to the FramedParts struct, but I fear using them in Framed::from_parts would be a subtle breaking change, and I am worried about having FramedRead::from_parts behave differently.

My preference would be to mirror the behaviour of Framed::from_parts, and maybe in tokio-util 0.8 we can reconsider. IIRC from_parts also seems to re-allocate if the buffers are deemed too small which upsets me as well.

Mrreadiness commented:
I didn't notice user can change the buffer via `read_buffer_mut()`. Strange that it's possible, but I agree - better default into `framing` state in such cases. Added this to MR.

> why capacity and not len?

I tried to look into history, but couldn't figure out the reasoning. In theory, it should be safe to replace `capacity` to `len` - you can't decode from empty buffer, right? But I'm still afraid to made this change. 

Is there anything else I can do to get this MR merged?
